hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/io/SmallFileDataStreamOutput.java

```diff
@@ -0,0 +1,564 @@
+    if (action.delayMillis > 0) {
+      try {
+        Thread.sleep(action.delayMillis);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        IOException ioe = (IOException) new InterruptedIOException(
+            "Interrupted: action=" + action + ", retry policy=" + retryPolicy)
+            .initCause(e);
+        setExceptionAndThrow(ioe);
+      }
+    }
+    retryCount++;
+    if (LOG.isTraceEnabled()) {
+      LOG.trace("Retrying Write request. Already tried {} time(s); " +
+          "retry policy is {} ", retryCount, retryPolicy);
+    }
+    //handleWrite();
+  }
+
+  /**
+   * Checks if the provided exception signifies retry failure in ratis client.
+   * In case of retry failure, ratis client throws RaftRetryFailureException
+   * and all succeeding operations are failed with AlreadyClosedException.
+   */
+  private boolean checkForRetryFailure(Throwable t) {
+    return t instanceof RaftRetryFailureException
+        || t instanceof AlreadyClosedException;
+  }
+
+  // Every container specific exception from datatnode will be seen as
+  // StorageContainerException
+  private boolean checkIfContainerToExclude(Throwable t) {
+    return t instanceof StorageContainerException;
+  }
+
+  private void cleanup(boolean invalidateClient) throws IOException {
+    if (xceiverClientFactory != null) {
+      xceiverClientFactory.releaseClient(xceiverClient, invalidateClient);
+    }
+    if (dataStreamOutput != null) {
+      try {
+        dataStreamOutput.close();
+      } catch (Exception e) {
+        throw new IOException("close dataStreamOutput error", e);
+      }
+    }
+    dataStreamOutput = null;
+    xceiverClient = null;
+  }
+
+  private ContainerProtos.ContainerCommandRequestProto getPutSmallFileRequest(
+      byte[] b) throws IOException {
+
+    ContainerProtos.BlockData containerBlockData =
+        ContainerProtos.BlockData.newBuilder()
+            .setBlockID(blockID.get().getDatanodeBlockIDProtobuf())
+            .build();
+    ContainerProtos.PutBlockRequestProto.Builder createBlockRequest =
+        ContainerProtos.PutBlockRequestProto.newBuilder()
+            .setBlockData(containerBlockData);
+
+    int size = b.length;
+
+    ByteString bytes = ByteString.copyFrom(b);
+
+    Checksum checksum =
+        new Checksum(config.getChecksumType(), config.getBytesPerChecksum());
+    final ChecksumData checksumData = checksum.computeChecksum(b);
+
+    ContainerProtos.ChunkInfo chunk =
+        ContainerProtos.ChunkInfo.newBuilder()
+            .setChunkName(blockID.get().getLocalID() + "_chunk_0")
+            .setOffset(0)
+            .setLen(size)
+            .setChecksumData(checksumData.getProtoBufMessage())
+            .build();
+
+    ContainerProtos.PutSmallFileRequestProto putSmallFileRequest =
+        ContainerProtos.PutSmallFileRequestProto.newBuilder()
+            .setChunkInfo(chunk)
+            .setBlock(createBlockRequest)
+            .setData(bytes)
```

Comment:
Do not set the data inside the proto.  We should send only the header in the proto and then send the raw data to the stream.